service/switchboardapp/src/test/java/com/switchboard/app/DecoderControllerTest.java

```diff
@@ -0,0 +1,137 @@
+package com.switchboard.app;
+
+import com.switchboard.app.controller.DecoderController;
+import com.switchboard.app.dao.DecoderDaoImpl;
+import com.switchboard.app.dao.DeviceDaoImpl;
+import com.switchboard.app.dto.DecoderDTO;
+import com.switchboard.app.dto.mapper.DecoderMapper;
+import com.switchboard.app.dto.mapper.DecoderMapperImpl;
+import com.switchboard.app.entity.DecoderEntity;
+import com.switchboard.app.entity.DeviceEntity;
+import com.switchboard.app.exceptions.ExceptionType.DeviceNotFoundException;
+import com.switchboard.app.fixture.DecoderFixture;
+import com.switchboard.app.fixture.DeviceFixture;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.*;
+import org.springframework.hateoas.EntityModel;
+import org.springframework.http.ResponseEntity;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.when;
+
+class DecoderControllerTest {
+
+    @InjectMocks
+    private DecoderController decoderController;
+
+    @Mock
+    private DecoderDaoImpl decoderService;
+
+    @Mock
+    private DeviceDaoImpl deviceService;
+
+    private DecoderMapper decoderMapper;
+
+    //stubbed Objects
+    static private DeviceEntity device1, device2;
+    static private DecoderEntity decoder1, decoder2;
+    static private List<DecoderEntity> listOfDecoders;
+
+    @BeforeAll
+    static void decoderFixture(){
+        //stubbing device and decoder objects
+        device1 = DeviceFixture.getDevice1();
+        device2 = DeviceFixture.getDevice2();
+        decoder1 = DecoderFixture.getDecoder1(device1);
+        decoder2 = DecoderFixture.getDecoder2(device2);
+        listOfDecoders = DecoderFixture.getListOfDecoders(decoder1, decoder2);
+    }
+
+    @BeforeEach
+    void setup(){
+        decoderMapper = Mockito.spy(new DecoderMapperImpl()); //to spy on DecoderMapper object
+        MockitoAnnotations.initMocks(this); //to be able to initiate decoderController object
+    }
+
+    @Test
+    final void testRetrieveAllDecoders(){
+        when(decoderService.getDecoders()).thenReturn(listOfDecoders);
+
+        List<DecoderDTO> allDecoders = decoderController.retrieveAllDecoders();
+        List<DecoderDTO> listOfExpectDTODecoders = decoderMapper.toDecoderDTOs(listOfDecoders); //covert List<DecoderEntity> to List<DecoderDTO>
```

Comment:
Ideally we would have another fixture method that returns a List<DecoderDTO>, but it's not a huge issue for now.